### PR TITLE
Managed appender

### PR DESCRIPTION
**Pull request by arontsang**

### CODE REVIEW

File: DuckDB.NET.Data/DuckDBAppender.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+public class DuckDBAppenderRow : IDisposable
+{
+	private readonly NET.DuckDBAppender _appender;
+
+	internal DuckDBAppenderRow(NET.DuckDBAppender appender)
+	{
+		_appender = appender;
+	}
+
+	public void AppendValue(bool value)
```
Comment (Giorgi) [inline DuckDB.NET.Data/DuckDBAppender.cs:61]: I think all `AppendXXX` methods should `return this` to have fluent interface.

File: DuckDB.NET.Data/DuckDBAppender.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	private readonly DuckDBConnection _connection;
+	private readonly string _table;
+	private readonly NET.DuckDBAppender _nativeAppender;
+
+	internal DuckDBAppender(DuckDBConnection connection, string table)
+	{
+		_connection = connection;
+		_table = table;
+
+		NativeMethods.Appender.DuckDBAppenderCreate(connection.NativeConnection, null, table, out _nativeAppender);
```
Comment (Giorgi) [inline DuckDB.NET.Data/DuckDBAppender.cs:16]: You should throw an exception if the return value is an error.

**Review by Giorgi — CHANGES_REQUESTED** [review]
Nice, I was also thinking about creating a managed wrapper around the appended API. I added some comments that need to be handled. Also, let's discuss some other necessary changes at [Discord .Net Channel](https://discord.com/channels/909674491309850675/1051088721996427265)

File: DuckDB.NET.Data/DuckDBAppender.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		NativeMethods.Appender.DuckDBAppendTime(_appender, NativeMethods.DateTime.DuckDBToTime(time));
+	}
+	
+	public void AppendValue(DuckDBTime time)
+	{
+		NativeMethods.Appender.DuckDBAppendTime(_appender, time);
+	}
+
+	#endregion
+
```
Comment (Giorgi) [inline DuckDB.NET.Data/DuckDBAppender.cs:176]: A separate method is required to append null values.

Comment (arontsang) [reply DuckDB.NET.Data/DuckDBAppender.cs:176]: Actually was thinking of overloading with nullable methods.

File: DuckDB.NET.Data/DuckDBAppender.cs (hunk 1 above)
Comment (arontsang) [reply DuckDB.NET.Data/DuckDBAppender.cs:61]: Fair enough

File: DuckDB.NET.Data/DuckDBAppender.cs (hunk 3 above)
Comment (Giorgi) [reply DuckDB.NET.Data/DuckDBAppender.cs:176]: That's also possible but I think having one method is easier to implement and better from user perspective too.

**Comment by arontsang** [comment]
Disagree from a "pit of success" point of view. The user would need to know
to call the null append method as opposed to the value append method.

Problem is worse with string/varchar. Nothing stops a user passing a null
string in...

**Comment by Giorgi** [comment]
@arontsang What do you think about having only nullable methods for `AppendValue` ? Does having nullable and non-nullable overloads provide any benefit?

**Comment by arontsang** [comment]
That is a very good point. Certainly, it would reduce complexity and
maintenance.

If however the point of the Appender being to be the most optimal solution
to insert rows into DuckDB, the non-nullable overloads would provide a tiny
perf boost.

Perhaps some benchmarking is in order.

**Comment by arontsang** [comment]
I actually didn't think about that at all when writing the PR...tree,
leaves and all that 😅

**Pull request merged**